app/main.py

```diff
@@ -1,34 +1,85 @@
 
 
 class Ship:
-    def __init__(self, start, end, is_drowned=False):
-        # Create decks and save them to a list `self.decks`
-        pass
+    def __init__(self,
+                 start: tuple[int, int],
+                 end: tuple[int, int],
+                 is_drowned: bool = False) -> None:
+        self.start = start
+        self.end = end
+        self.is_drowned = is_drowned
+        self.decks = []
+
+        if start[0] == end[0]:
+            for i in range(min(start[1], end[1]), max(start[1], end[1]) + 1):
+                self.decks.append(Deck(start[0], i))
+        elif start[1] == end[1]:
+            for i in range(min(start[0], end[0]), max(start[0], end[0]) + 1):
+                self.decks.append(Deck(i, start[1]))
+
+    def get_deck(self, row: int, column: int) -> Deck:
+        return next((deck for deck in self.decks if
+                     deck.row == row and deck.column == column), None)
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def fire(self, row: int, column: int) -> str:
+        if self.is_drowned:
+            return "Already Sunk!"
+        current_deck = self.get_deck(row, column)
+        if not current_deck.is_alive:
+            return "Already hit!"
+        current_deck.is_alive = False
+        if not any(deck.is_alive for deck in self.decks):
+            self.is_drowned = True
+            return "Sunk!"
+        return "Hit!"
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def __repr__(self) -> str:
+        return (f"Ship from ({self.start[0]}, {self.start[1]})"
+                f" to ({self.end[0]}, {self.end[1]})")
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: Iterable) -> None:
+        self.field = {}
+        self.ships = []
+        self.area_size = 10
+        self.area = [["   ~   " for _ in range(self.area_size)] for _ in
+                     range(self.area_size)]
+
+        for ship_coords in ships:
+            ship = Ship(ship_coords[0], ship_coords[1])
+            self.ships.append(ship)
+            for deck in ship.decks:
+                if (deck.row, deck.column) in self.field:
+                    raise ValueError(
+                        f"Invalid ship placement: "
+                        f"overlap at ({deck.row}, {deck.column})")
```

Comment:
The error message for overlapping ships is correct, but it could be more informative by including more context, such as the coordinates of both overlapping ships.